addons/portal/controllers/portal.py

```diff
@@ -167,6 +186,100 @@ def _prepare_portal_overview_values(self):
             }]
         }
 
+    def _is_own_partner_address(self, partner, address_id):
+        """Checks whether address_id is the id of a res.partner representing:
+            - the main address of partner (partner.id)
+            - an invoice address from partner.child_ids
+            - a delivery address from partner.child_ids
+        """
+        # if address_id is partner's main address
+        if partner.id == address_id:
+            return True
+        # if address_id partner's secondary address
+        return address_id in partner.child_ids.filtered(lambda partner: partner.type in ['invoice', 'delivery']).ids
+
+    def _render_portal_my_address(self, values):
+        response = request.render("portal.portal_my_address", values)
+        response.headers['X-Frame-Options'] = 'DENY'
+        return response
+
+    def _prepare_address_form_values(self, address_id=None):
+        values = self._prepare_portal_layout_values()
+
+        partner = request.env.user.partner_id
+        address = request.env['res.partner'].with_context(show_address=1).browse(address_id) if address_id else False
+
+        if address:
+            fields = {**self.MANDATORY_PARTNER_FIELDS, **self.OPTIONAL_PARTNER_FIELDS}
+            # add address data
+            values.update({value: getattr(address, key) for key, value in fields.items()})
+
+        countries = request.env['res.country'].sudo().search([])
+        states = request.env['res.country.state'].sudo().search([])
```

Comment:
suffix by _sudo so we can easily see this is sudo'ed later in the code :)
EDIT: well, this could be inlined inside the values.update()

Also, do we really need all the countries and states of the entire terrestrial planet ?
If i understand well, this is used to allow the user to select his country and state when he edit/add an address. Maybe only give the countries and after the country has been selected, search the states by rpc using the given country.

Reply:
Are we using select2 ? It effectively generally makes an RPC to fetch data to display ... will see when arriving on its usage ^^ .